Generate dpkg.yaml

ilasc opened the pull request:
Proposed branch to create the dpkg.yaml file needed for the OSS compliance project (per discussions in design doc "LP101 - Minimal support for OSS compliance on selected snaps").

Commit:
Generate dpkg.yaml

ilasc commented:
Tested by building locally, dpkg.yaml generated during build: https://pastebin.canonical.com/p/zw8mqDvZXH/

sil2100 reviewed (dismissed):
Ok, I think this looks decent from an initial look. So I'm +1 on the changes.

But I have one request: could you please add a simple test case for it into `hook-tests/`? It feels like something we could test, e.g. mocking the addition of a ppa, checking if the resulting yaml file has it. Mock a `dpkg.list` and checking if expected packages are present. Thanks!

Commit:
Add test for 602-create-dpkg-yaml.chroot

stulluk commented:
Thank you for the test, but if I understand correctly, if there is no PPA section, basically test will fail ?

I am asking this because for future base snaps ( core22, core24 maybe ) we want to get rid of PPA at all ( everything should come from the ubuntu archive )

Also, I am not %100 sure about comparing string contents of dpkg packages is the correct way of doing this test ? Is there a possibility that a version of a package in 18.04 base tarball to change for example ?

ilasc commented:
Thanks for the prompt feedback! 

That is correct, that test will fail if the yaml file doesn't contain that PPA. But it fails with a suggestion to what needs to change - I was trying to follow the procedure you're following currently for other hooks / tests (for example 001-extra-packages.test). 

This approach will work for future base snaps ( core22, core24 ) even if you will eliminate all PPAs. From what I've seen in the way you are coding the snaps: each core snap has its own set of hooks and its separate set of tests, currently the hook I proposed for Core20 differs from the one I had to propose for Core18. So for core22 and core24 each hook and corresponding test will have a structure that complies with the way you build these snaps (same as I implemented core core18 and core20).

I agree that the version is not a good idea. I will adjust to eliminate versions from the test in the next commit.

mvo5 commented on hook-tests/602-create-dpkg-yaml.test:
```diff
@@ -0,0 +1,256 @@
+#third check the Packages section of dpkg.yaml
+packages_section=$(cat <<EOF
+packages:
+- adduser=3.116ubuntu1
```
Adding a test that checks for these exact package versions seems risky (unless I miss something) - the core18 snap will get new versions because of e.g. security updates and then this test will start failing

mvo5 commented on hooks/602-create-dpkg-yaml.chroot:
```diff
@@ -0,0 +1,25 @@
+  
+# fill in yaml section with all installed packages
+  printf 'packages:\n'
+  tail -n +6 /usr/share/snappy/dpkg.list | awk '{printf "- "}{printf $2}{printf "="}{print $3}'
```
Please also see the comments for https://github.com/snapcore/core/pull/121#discussion_r692182036

mvo5 reviewed (dismissed):
Thanks for the PR and the tests. Please check my comments there, unless I miss soemthing the current approach is a bit too brittle. Overall it looks great though. Just marking as "need chagnes" to ensure it does not get merged.

Commit:
Add test for generating dpkg.yaml

ilasc commented:
Removed package versions from tests and in the generation of the yaml file I removed hardcoding of the point where the packages section begins in dpkg.list.

Commit:
Move generation of dpkg.yaml to 600-no-debian.chroot

ilasc commented:
Following changes discussed for the Core Snap (snapcore/core#121) now using dpkg-query to generate the packages section in the yaml file and removed package versions from the test.

mvo5 commented on hook-tests/600-no-debian.test:
```diff
@@ -1,4 +1,4 @@
-#!/bin/sh -e
+#!/bin/sh
```
Why is the `set -e` removed here?

ilasc replied:
Well spotted Michael, that's added back in now with the latest commit.

mvo5 commented on hook-tests/600-no-debian.test:
```diff
@@ -24,14 +24,13 @@ else
   exit 1
 fi
 
-#third check the Packages section of dpkg.yaml
```
Why is the comment removed?

ilasc replied:
That's added back in now with the latest commit.

mvo5 commented on hook-tests/600-no-debian.test:
```diff
@@ -265,6 +263,6 @@ else
 fi
 
-# cleanup temp files used for comparisons 
+# cleanup temp files used foe comparisons 
```
This looks like a typo

ilasc replied:
Indeed, that's corrected now with the latest commit.

mvo5 commented on hooks/600-no-debian.chroot:
```diff
@@ -10,6 +10,25 @@ echo "I: Removing the debian legacy"
+  
+  # fill in yaml section with all installed packages
+  printf 'packages:\n'
+  dpkg-query -W --showformat='- ${binary:Package}=${Version}\n'
```
Thanks for this, this looks nice!

xnox replied:
do we need any quoting here? i'm pondering if the combination of debian binary package name = debian valid versions are all safe literals for yaml.

ilasc replied:
@xnox thanks for the feedback , this produces a valid yaml file.

mvo5 reviewed (commented):
Thanks for the update, some questions inline about the changes in the test code, the actual functionality change looks very nice (the dpkg-query -W)

Commit:
Cleanup typos and omitted test code

sil2100 commented:
Left some comments on https://github.com/snapcore/core20/pull/108#pullrequestreview-738301829 <- maybe some of those are relevant here as well?

stulluk commented:
> Left some comments on [snapcore/core20#108 (review)](https://github.com/snapcore/core20/pull/108#pullrequestreview-738301829) <- maybe some of those are relevant here as well?

When all these merged, I want to remind core22 repository as well.

Commit:
Simplify test for 600-no-debian.chroot

Commit:
Remove hooks/602-create-dpkg-yaml.chroot

sil2100 reviewed (approved):
Ok, this looks similar to the core20 PR - looking good. Did a local test build and confirmed the tests are passing.

stulluk reviewed (approved):
LGTM

mvo5 reviewed (approved):
Looks very nice now, thank you
